pkg/controller/storagecluster/cephcluster.go

```diff
@@ -64,7 +64,11 @@ func (r *ReconcileStorageCluster) ensureCephCluster(sc *ocsv1.StorageCluster, re
 	if sc.Spec.ExternalStorage.Enable {
 		cephCluster = newExternalCephCluster(sc, r.cephImage, r.monitoringIP)
 	} else {
-		cephCluster = newCephCluster(sc, r.cephImage, r.nodeCount, r.serverVersion, reqLogger)
+		var err error
+		cephCluster, err = r.newCephClusterWithKMS(sc, r.cephImage, r.nodeCount, r.serverVersion, reqLogger)
```

Comment:
I tried different scenarios and finally arrived at this change. The main difficulty is that we have to change the existing functional type of the current `newCephCluster` function (which is being used in multiple places, including in different tests).

For the above requirement, TWO main changes are needed to the existing function,

a. We require an instance of `ReconcileStorageCluster`, which is needed to retrieve KMS details. So, either we have to pass it as an argument or add this function as a method of the `ReconcileStorageCluster` object. Both of which changes the existing functional type of `newCephCluster`.

b. Once we decide to add/attach KMS details to the CephCluster, we have to return the error value as well along with `CephCluster` object, which again changes functional type.

In the current way, the existing function is not changed. If user / developer wants a KMS attached CephCluster they have to use `ReconcileStorageCluster`'s method `newCephClusterWithKMS`.

Reply:
> I tried different scenarios and finally arrived at this change. The main difficulty is that we have to change the existing functional type of the current `newCephCluster` function (which is being used in multiple places, including in different tests).
>
Honestly, I'd be fine with that. Specially, if it means improved code readability and clean & clear code.
 
> For the above requirement, TWO main changes are needed to the existing function,
> 
> a. We require an instance of `ReconcileStorageCluster`, which is needed to retrieve KMS details. So, either we have to pass it as an argument or add this function as a method of the `ReconcileStorageCluster` object. Both of which changes the existing functional type of `newCephCluster`.
> 
> b. Once we decide to add/attach KMS details to the CephCluster, we have to return the error value as well along with `CephCluster` object, which again changes functional type.
> 
> In the current way, the existing function is not changed. If user / developer wants a KMS attached CephCluster they have to use `ReconcileStorageCluster`'s method `newCephClusterWithKMS`.
>
I see what you mean. But, still I'm not very comfortable with multiple generator functions for the same object. And, `KMS` defines behaviour of `CephCluster` and not `ReconcileStorageCluster`, so I would prefer not to have `r.newCephClusterWithKMS()` or `r.addKmsToCephCluster()`.

Can we do this instead:
We retain `newCephCluster()` that returns a basic `CephCluster` object. We then add a method to modify it's behaviour i.e `CephCluster.addKMS(client r.client)` and the call to this modifier will be safeguarded which shouldn't be that bad as we'll do it once. 

So code would look something like
```
cephCluster = newCephCluster(sc, r.cephImage, r.nodeCount, r.serverVersion, reqLogger)
if sc.Spec.Encryption.KeyManagementService.Enable{
 err := cephCluster.addKMS(r.client, ...)
  if err != nil{
     //reconcile
  }
}
```
 Wouldn't this be more manageable, read better and share less data?

Reply:
> I see what you mean. But, still I'm not very comfortable with multiple generator functions for the same object.

This is a common scenario in `golang` where you use different functions to initialize an object in different ways.
Golang's [context](https://golang.org/pkg/context/) package is an example, where we have functions like[
WithCancel](https://golang.org/pkg/context/#WithCancel) , [WithDeadline](https://golang.org/pkg/context/#WithDeadline), etc functions which returns different `Context`s according to the input/argument `Context`.
More such functions could be found with `New***()` names, where objects are created according to different arguments passed (something like `NewBuffer(buf []byte) *Buffer` ,  `NewBufferString(s string) *Buffer` etc)

> And, KMS defines behaviour of CephCluster and not ReconcileStorageCluster, so I would prefer not to have r.newCephClusterWithKMS() or r.addKmsToCephCluster().

I totally agree with you.

> We retain newCephCluster() that returns a basic CephCluster object. We then add a method to modify it's behaviour i.e CephCluster.addKMS(client r.client) and the call to this modifier will be safeguarded which shouldn't be that bad as we'll do it once.


`CephCluster` is an object from `github.com/rook/rook/pkg/apis/ceph.rook.io/v1` (nicknamed `cephv1`) and we won't be able to add functions into it directly from our operator code base. Or we have to create a type alias and add the function to it, which complicates the matter.
Another question I have is, if we already know that we require a `CephCluster` with the KMS details, why not make both the lines (creation of cephcluster and the safeguard `if` condition following it) together into a single function and just call it. This will be a much safer approach and this function can be reused in places where we require a `CephCluster` with KMS details (let's say, in writting test cases).

If you are convinced about my first point (about different functions being used to initialize same object and it is a common practice in `golang` world), we can remove all the `ReconcileStorageCluster` methods (that is `r.methods`) and put them as standard methods. So there will be TWO methods to create converged mode `CephCluster`.

a. `newCephCluster` : the existing method without any change
b. `newCephClusterWithKMS`: a new standalone method (like the above one) with extra `client` parameter passed. Along with `CephCluster`, this will return an additional error value.


If you are still not convinced, we have to go with the longer approach of changing the existing `newCephCluster` function (which will now take a `client` and returns an additional error). The con/down-side with this approach is all the test files, which uses this method, have to be changed and additionally we won't be able to create a ceph cluster (which does not have KMS enabled) without using a `client`.

My recommendation will be to add two methods instead of one.

Reply:
I have updated PR (on the agreed point that we should remove the methods from the `ReconcileStorageCluster` object and made them standalone).